controllers/persistentvolumeclaim_controller.go

```diff
@@ -0,0 +1,252 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// PersistentVolumeClaimReconciler reconciles a PersistentVolumeClaim object
+type PersistentVolumeClaimReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+var (
+	rsCronJobScheduleTimeAnnotation = "reclaimspace." + csiaddonsv1alpha1.GroupVersion.Group + "/schedule"
+	rsCronJobNameAnnotation         = "reclaimspace." + csiaddonsv1alpha1.GroupVersion.Group + "/cronjob"
+)
+
+const (
+	defaultSchedule = "@weekly"
+)
+
+//+kubebuilder:rbac:groups=core,resources=persistentvolumeclaims,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core,resources=persistentvolumeclaims/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=core,resources=persistentvolumeclaims/finalizers,verbs=update
+//+kubebuilder:rbac:groups=csiaddons.openshift.io,resources=reclaimspacecronjobs,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// This is triggered when `reclaimspace.csiaddons.openshift/schedule` annotation
+// is found on newly created PVC or if there is a change in valude of the annotation.
+func (r *PersistentVolumeClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	// Fetch PersistentVolumeClaim instance
+	pvc := &corev1.PersistentVolumeClaim{}
+	err := r.Client.Get(ctx, req.NamespacedName, pvc)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			logger.Info("PersistentVolumeClaim resource not found")
+
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	// delete any old child cronjobs.
+	err = r.deleteChildCronJobs(ctx, &logger, &req)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	annotations := pvc.GetAnnotations()
+	schedule, scheduleFound := getScheduleFromAnnotation(&logger, annotations)
+	if !scheduleFound {
+		// if schedule is not found and name is found, delete name from annotation.
+		_, nameFound := annotations[rsCronJobNameAnnotation]
+		if nameFound {
+			delete(annotations, rsCronJobNameAnnotation)
+			pvc.SetAnnotations(annotations)
+			err = r.Client.Update(ctx, pvc)
+			if err != nil {
+				logger.Error(err, "Failed to update annotation")
+
+				return ctrl.Result{}, err
+			}
+		}
+		logger.Info("Annotation not set, exiting reconcile")
+		// no schedule annotation set, just dequeue.
+		return ctrl.Result{}, nil
+	}
+
+	// create cronjob name and update in annotations.
+	rsCronJobName := fmt.Sprintf("%s-%v", pvc.Name, time.Now().Unix())
+	annotations[rsCronJobNameAnnotation] = rsCronJobName
+	pvc.SetAnnotations(annotations)
+	err = r.Client.Update(ctx, pvc)
+	if err != nil {
+		logger.Error(err, "Failed to update annotation")
+
+		return ctrl.Result{}, err
+	}
```

Comment:
set annotation only after creating the cronJob?

Reply:
done